File: pca.R
library(ggbiplot)

# LOAD THE DATA
data <- read.table('NC_001416.aa', sep='\t', header=TRUE)

# STRIP OUT ONLY THE COUNTS
dat <- data[,unlist(strsplit("ARNDCEQGHILKMFPSTWYV", ""))]

# DIVIDE BY THE ROW SUMS TO GET THE FREQUENCY
dat.norm <- dat/rowSums(dat)

# SCALE THE VALUES
dat.scaled <- scale(dat.norm)

# CLASSIFY EACH ROW USING KMEANS
clust <- kmeans(dat.scaled, 2)$cluster

# CALCULATED THE PCA
dat.pca <- prcomp(dat.scaled, center = FALSE, scale. = FALSE)

# PLOT
g <- ggbiplot(dat.pca, 
			  groups = as.factor(data$TYPE), 
			  varname.size = 5, 
			  obs.scale = 3, 
			  var.scale = 1, 
			  ellipse = TRUE, 
			  circle = FALSE, 
			  alpha = 1
			  )
# LEGEND
g <- g + theme(legend.direction = 'vertical', 
	legend.justification=c(0,1), 
	legend.position=c(0.8, 0.95),
	legend.background = element_rect(linetype = 2, size = 0.5, colour = 1)),
	legend.key = element_blank()
	)
# OPTIONS
g <- g + geom_point(aes(colour=as.factor(clust), shape=as.factor(clust)), size=3, pch=22, alpha=1 )
g <- g + ggtitle("amino-acid frequency of potential ORFs from Lambda phage")
g <- g + theme(plot.title = element_text(hjust = 0.5))
g <- g + guides(color=guide_legend(title="TYPE")) 
#g <- g + scale_color_manual(name="TYPE", values=c("blue", "red", "blue", 'red'))
#g <- g + xlim(-6, 6) + ylim(-6, 6)
#print(g)

# CREATE TIFF IMAGE
tiff(filename="figure.tiff", width=8, height=6, units="in", res=200)
print(g)
invisible(dev.off())
cat("Output image: figure.tiff\n")

File: pca.py
import pandas as pd
from sklearn.preprocessing import StandardScaler
#from sklearn.cluster import KMeans
from sklearn import decomposition
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

# LOAD THE DATA
data = pd.read_csv('NC_001416.aa', sep='\t')

# STRIP OUT ONLY THE COUNTS
dat = data[list("ARNDCEQGHILKMFPSTWYV")]

# DIVIDE BY THE ROW SUMS TO GET THE FREQUENCY
dat_norm = dat.div(dat.sum(axis=1), axis=0)

# SCALE THE VALUES
dat_scaled = StandardScaler().fit_transform(dat_norm)

# CLASSIFY EACH ROW USING KMEANS
#clust = KMeans(n_clusters=2).fit(dat_scaled).labels_

# CALCULATE THE PRINCIPLE COMPONENTS
pca = decomposition.PCA(n_components = 2, svd_solver='full').fit(dat_scaled)
dat_pca = pca.transform(dat_scaled)

x_vector = pca.components_[0]
y_vector = pca.components_[1]

# PLOT
fig, ax = plt.subplots()
colors = {'noncoding':'#F2766E', 'coding':'#3CC9CF'}

df = pd.DataFrame({'X':dat_pca[:,0],'Y':dat_pca[:,1],'TYPE':data.TYPE})

ax.scatter(df['X'], df['Y'], label=df['TYPE'], c=df['TYPE'].apply(lambda x: colors[x]), marker='.', linewidths=0.0, alpha=0.9, zorder=5)

for i in range(len(x_vector)):
    x = (1.2*x_vector[i]*max(dat_pca[:,0]))
    y = (1.2*y_vector[i]*max(dat_pca[:,0]))
    plt.arrow(0, 0, x, y,  color='black', width=0.00005, zorder=10)
    plt.text(x*1.1, y*1.1, dat.columns[i], color='black', zorder=10)

# LEGEND
plt.legend(handles=[mpatches.Patch(color=col, label=str(lab)) for lab,col in colors.items()])

ax.set_title('amino-acid frequency of potential ORFs from Lambda phage')
ax.set(xlabel='PC1', ylabel='PC2')
#plt.show()
fig.set_size_inches(8, 8)
fig.savefig('figure2.png', dpi=100, format='png')

